src/python/pants/backend/python/dependency_inference/rules_test.py

```diff
@@ -744,6 +744,40 @@ def test_infer_python_strict_multiple_resolves(imports_rule_runner: PythonRuleRu
         )
 
 
+def test_infer_python_identical_files_with_relative_imports_should_be_treated_differently(
+    imports_rule_runner: PythonRuleRunner,
+) -> None:
+    # dependency inference shouldn't cache _just_ based on file contents, because this can break
+    # relative imports. When b reused a's results, b/__init__.py was incorrectly depending on
+    # a/file.py
+    contents = "from . import file"
+    imports_rule_runner.write_files(
+        {
+            "a/BUILD": "python_sources()",
+            "a/__init__.py": contents,
+            "a/file.py": "",
+            "b/BUILD": "python_sources()",
+            "b/__init__.py": contents,
+            "b/file.py": "",
+        }
+    )
+
+    for directory in ["a", "b"]:
+        tgt = imports_rule_runner.get_target(
+            Address(directory, target_name=directory, relative_file_path="__init__.py")
+        )
+
+        deps = imports_rule_runner.request(
+            InferredDependencies,
+            [InferPythonImportDependencies(PythonImportDependenciesInferenceFieldSet.create(tgt))],
+        )
+        assert deps == InferredDependencies(
+            [
+                Address(directory, target_name=directory, relative_file_path="file.py"),
+            ]
+        )
```

Comment:
Instead of a loop, can we unroll this and hardcode the values so it's very clear that were:

- requesting "`a/__init__.py`
- Asserting it depends on `a/file.py`
- requesting `b/__init__.py`
- Asserting it depends on `b/file.py`
  - With a comment explaining that without special handling, this used to return `a/file.py`

Otherwise it'd be easy to "refactor" it to pass, but exhibit the bug if someone changes the `target_name=directory`.